compiler/rustc_resolve/src/imports.rs

```diff
@@ -1367,27 +1346,31 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
         }
     }
 
-    fn resolve_glob_import(&mut self, import: Import<'a>) {
+    fn resolve_glob_import(&mut self, import: Import<'a>) -> usize {
         // This function is only called for glob imports.
         let ImportKind::Glob { id, is_prelude, .. } = import.kind else { unreachable!() };
 
         let ModuleOrUniformRoot::Module(module) = import.imported_module.get().unwrap() else {
             self.tcx.sess.create_err(CannotGlobImportAllCrates { span: import.span }).emit();
-            return;
+            return 0;
         };
 
         if module.is_trait() {
             self.tcx.sess.create_err(ItemsInTraitsAreNotImportable { span: import.span }).emit();
-            return;
+            return 0;
         } else if module == import.parent_scope.module {
-            return;
+            return 0;
         } else if is_prelude {
             self.prelude = Some(module);
-            return;
+            return 0;
+        } else if module.unexpanded_invocations.borrow().len() > 0 {
+            return 3;
         }
 
         // Add to module's glob_importers
-        module.glob_importers.borrow_mut().push(import);
+        let prev = module.glob_importers.borrow_mut().insert(import);
+        // Ensure that it's inserted only once.
+        assert!(prev);
```

Comment:
I don't understand why this assert isn't hit, perhaps we are missing a test?

`resolve_import` can now return a non-zero for a glob import, so it can return into the `indeterminate_imports` queue and be resolved again later.

The logic behind the existing code is that `resolve_glob_import` is called only once for a glob import (when its module is initially resolved) and performs initial setup - does initial update and fills `glob_importers` for further updates.
All further updates then go through `update_resolution` when something in other modules changes.
It think this is the right setup, and don't think it needs to be changed.

Reply:
> It think this is the right setup, and don't think it needs to be changed.

Yes, I will revert it